internal/daemon/controller/handlers/workers/grants_test.go

```diff
@@ -58,76 +64,1821 @@ func TestGrants_ReadActions(t *testing.T) {
+			}),
+			wantOutputFields: []string{globals.IdField, globals.VersionField, globals.ConfigTagsField, globals.TypeField},
+			wantErr:          nil,
+		},
+		{
+			name: "global role grant allow different ID returns error",
+			input: &pbs.GetWorkerRequest{
+				Id: globalWorker1.PublicId,
+			},
+			userFunc: iam.TestUserManagedGroupGrantsFunc(t, conn, kmsCache, org.PublicId, ldap.TestAuthMethodWithAccountInManagedGroup, []iam.TestRoleGrantsRequest{
+				{
+					RoleScopeId: globals.GlobalPrefix,
+					Grants:      []string{fmt.Sprintf("ids=%s;actions=read", globalWorker2.PublicId)},
+					GrantScopes: []string{globals.GrantScopeThis},
+				},
+			}),
+			wantErr: handlers.ApiErrorWithCode(codes.PermissionDenied),
+		},
+		{
+			name: "global role grant different action returns error",
+			input: &pbs.GetWorkerRequest{
+				Id: globalWorker1.PublicId,
+			},
+			userFunc: iam.TestUserManagedGroupGrantsFunc(t, conn, kmsCache, org.PublicId, ldap.TestAuthMethodWithAccountInManagedGroup, []iam.TestRoleGrantsRequest{
+				{
+					RoleScopeId: globals.GlobalPrefix,
+					Grants:      []string{"ids=*;type=*;actions=set-worker-tags"},
+					GrantScopes: []string{globals.GrantScopeThis},
+				},
+			}),
+			wantErr: handlers.ApiErrorWithCode(codes.PermissionDenied),
+		},
+		{
+			name: "global role grant different resource returns error",
+			input: &pbs.GetWorkerRequest{
+				Id: globalWorker1.PublicId,
+			},
+			userFunc: iam.TestUserManagedGroupGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, ldap.TestAuthMethodWithAccountInManagedGroup, []iam.TestRoleGrantsRequest{
+				{
+					RoleScopeId: globals.GlobalPrefix,
+					Grants:      []string{"ids=*;type=group;actions=*"},
+					GrantScopes: []string{globals.GrantScopeThis},
+				},
+			}),
+			wantErr: handlers.ApiErrorWithCode(codes.PermissionDenied),
+		},
+		{
+			name: "global role grant descendants resource returns error",
+			input: &pbs.GetWorkerRequest{
+				Id: globalWorker1.PublicId,
+			},
+			userFunc: iam.TestUserManagedGroupGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, ldap.TestAuthMethodWithAccountInManagedGroup, []iam.TestRoleGrantsRequest{
+				{
+					RoleScopeId: globals.GlobalPrefix,
+					Grants:      []string{"ids=*;type=*;actions=*"},
+					GrantScopes: []string{globals.GrantScopeDescendants},
+				},
+			}),
+			wantErr: handlers.ApiErrorWithCode(codes.PermissionDenied),
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			user, accountID := tc.userFunc()
+			tok, err := atRepo.CreateAuthToken(ctx, user, accountID.GetPublicId())
+			require.NoError(t, err)
+			fullGrantAuthCtx := cauth.TestAuthContextFromToken(t, conn, wrapper, tok, iamRepo)
+			got, finalErr := s.GetWorker(fullGrantAuthCtx, tc.input)
+			if tc.wantErr != nil {
+				require.ErrorIs(t, finalErr, tc.wantErr)
+				return
+			}
+			require.NoError(t, finalErr)
+			// validate that we're getting the right ID back
+			require.Equal(t, tc.input.Id, got.GetItem().GetId())
```

Comment:
Add output field assertion -

```suggestion
			require.Equal(t, tc.input.Id, got.GetItem().GetId())
            handlers.TestAssertOutputFields(t, got.GetItem(), tc.wantOutputFields)
```

Reply:
oops 😅  good catch!